algebra/src/curves/models/short_weierstrass_jacobian.rs

```diff
@@ -277,6 +277,77 @@ impl<P: Parameters> AffineCurve for GroupAffine<P> {
     fn mul_by_cofactor_inv(&self) -> Self {
         self.mul(P::COFACTOR_INV).into()
     }
+
+    fn apply_endomorphism(&self) -> Self { 
+        let mut self_e = self.clone();
+        self_e.x.mul_assign(P::ENDO_COEFF);
+        self_e
+    }
+
+    fn endo_rep_to_scalar(bits: Vec<bool>) -> Self::ScalarField {
+
+        let mut a : P::ScalarField = 2u64.into();
+        let mut b : P::ScalarField = 2u64.into();
+
+        let one = P::ScalarField::one();
+        let one_neg = one.neg();
+
+        for i in (0..(bits.len() / 2)).rev() {
+            a.double_in_place();
+            b.double_in_place();            
+
+            let s =
+                if bits[i * 2] {
+                    &one
+                } else {
+                    &one_neg
+                };
+
+            if bits[i * 2 + 1] {
+                a.add_assign(s);
+            } else {
+                b.add_assign(s);
+            }
+        }
+
+        a.mul(P::ENDO_SCALAR) + &b
+    }
+
+    /// Performs scalar multiplication of this element with mixed addition.
+    fn endo_mul(&self, bits: Vec<bool>) -> Self::Projective {
+
+        let self_neg = self.neg();
+
+        let self_e = self.apply_endomorphism();
+        let self_e_neg = self_e.neg();
+
+        let mut acc = self_e.into_projective();
+        acc.add_assign_mixed(&self);
+        acc.double_in_place();
+
+        for i in (0..(bits.len() / 2)).rev() {
+
+            let s =
+                if bits[i * 2 + 1] {
+                    if bits[i * 2] {
+                        &self_e
+                    } else {
+                        &self_e_neg
+                    }
+                } else {
+                    if bits[i * 2] {
+                        &self
+                    } else {
+                        &self_neg
+                    }
+                };
+
+            acc.double_in_place();
+            acc.add_assign_mixed(s);
+        }
+
+        acc
+    }
```

Comment:
@DDT92  In primitive, is the endo mul in fact faster than double and add on the equivalent scalar (i.e. the default implementation for the `AffineCurve` trait)?

Reply:
I assume that by counts of doublings and adds, it is.